### ISSUE TITLE
Postgres JSONB column is returned as string instead of hash in Rails 5 RC2

### ISSUE THREAD

**Issue by drewhamlett** [rails/rails#25594]
### Steps to reproduce
Running Rails 4.2.4.
Upgraded to Rails 5 RC2
Dropped and recreated the database.
All tests that use JSONB column are failing because column is not being serialized into a hash anymore.  Coming back as a plain string.  I have been unable to find any information on the internet about this.  schema.rb file shows that the column is jsonb.

## Edit
I looked at this issue some more.  It seems that using quotes in the migration script allows the default value to be created in Postgres as a string.

In Rails 4 we had this.  Which allowed the column to be created in Postgres correctly without quotes.
```rb
class CoolMigration < ActiveRecord::Migration
  def change
    add_column :businesses, :feature_flags, :jsonb, default: '{}'
  end
end
```
If I run this same migration with Rails 5 it creates the default value in Postgres as "{}".  This causes Rails to serialize this out into a string.  If change the migration to 

```rb
add_column :businesses, :feature_flags, :jsonb, default: {}
```

Everything works fine

### Expected behavior
Postgres jsonb column should be serialized to hash

### Actual behavior
Postgres jsonb column comes back as a string

### System configuration
**Rails version**: 5.0.0.rc2

**Ruby version**: 2.3.0

**Comment by sgrif** [comment]
This is the expected behavior. `"{}"` is a valid JSON value.

**Comment by drewhamlett** [comment]
Hey @sgrif Thanks for getting back.  We have noticed that there is a regression from Rails 4 to 5 with migrations.

These behave differently in Rails 5, where is in Rails 4 they all get set to {} in Postgres.
```rb
add_column :users, :meta, :jsonb, default: '{}'
add_column :users, :meta2, :jsonb, default: {}
```

**Comment by sgrif** [comment]
The behavior of

```ruby
add_column :users, :meta, :jsonb, default: '{}'
```

in 4.2 was a bug, not intended behavior.

**Comment by drewhamlett** [comment]
Ok.  Thanks for your help!

**Comment by sgrif** [comment]
See efaa6e4f for more reasoning

**Comment by nickpoorman** [comment]
This should probably be mentioned somewhere in the upgrade guide as most of the tutorials out there tell you to use the string version:

e.g.`add_column :users, :meta, :jsonb, default: '{}'`.

There's going to be a lot of people wondering why their migrations are breaking.

**Comment by sgrif** [comment]
Yes, it should have been added. Feel free to open a PR if not.

**Comment by drewhamlett** [comment]
@nickpoorman Do you want to make a PR for this?

**Comment by biznickman** [comment]
@drewhamlett @sgrif So what's the "expected behavior" here? Do you need to manually parse all values returned from a Postgres :jsonb field?

**Comment by sgrif** [comment]
@biznickman No. `default: {}` will give you an empty hash. `default: "{}"` will give you a string containing curlies. Both are valid JSON primitives, and Rails will give you whichever you specify.

**Comment by jerefrer** [comment]
I just followed [this tutorial](http://nandovieira.com/using-postgresql-and-jsonb-with-ruby-on-rails) to start using `jsonb` in my Rails 5.0.0 app.

After doing what was explained I had weird issues telling me that a `String` was not expected.
(`NoMethodError: undefined method `with_indifferent_access' for "{}":String`)

I then googled `Rails 5 jsonb` and found this issue which solved my problem by making me switch from `default: "{}"` to `default: {}`.

I checked and it doesn't seem to be anything about this in the upgrading guide.
Any news on this ?

**Comment by maclover7** [comment]
@jerefrer Feel free to submit a PR updating the guide.

**Comment by MCodyB** [comment]
NB: if you check your schema and it's a hash literal that doesn't necessarily mean anything, check the migration.

**Comment by emaillenin** [comment]
While saving to the database, save it as a hash. When you are reading it back, you will get back a hash. Eg:

````
Model.create(payload: JSON.parse(params[:payload]))
````

**Comment by TangMonk** [comment]
```
add_column :businesses, :feature_flags, :jsonb, default: {}
```

not works at rails 5.1.1

**Comment by marcelokanzaki** [comment]
For anyone looking to migrate the old way:

```ruby
class ChangeMyTableColumnDefault < ActiveRecord::Migration[5.0]
  def up
    change_column_default :my_table, :column, from: '{}', to: {}
    execute "UPDATE my_table SET column = '{}'::jsonb WHERE column = '\"{}\"'"
  end

  def down
    change_column_default :my_table, :column, from: {}, to: '{}'
    execute "UPDATE my_table SET column = '\"{}\"' WHERE column = '{}'::jsonb"
  end
end
```

**Comment by rromanchuk** [comment]
@Frank004 same. Not sure how to fix 

```ruby
class JSONHashSerializer
  def self.dump(hash)
    hash.to_json
  end

  def self.load(hash)
    (hash || {}).with_indifferent_access
  end
end

```

**Comment by singjsong** [comment]
@Frank004 @rromanchuk changing `load` to the following should be Good Enough™:
```ruby
hash ? JSON.parse(hash, symbolize_names: true) : {}
```
I'm only looking for a quick dirty fix myself and this is it. It seems though that the more proper way of addressing this is using the new Attributes API. This guy [did it](http://til.obiefernandez.com/posts/8c31a92080-rails-5-attributes-api-jsonb-postgres-columns).

**Comment by singjsong** [comment]
@Frank004 @rromanchuk changing `load` to the following seems to be good enough:
```ruby
if hash.nil?
  {}
elsif hash.is_a?(Hash) && hash.empty?
    hash
else
    JSON.parse(hash)
end.with_indifferent_access
```
I'm just looking for a quick and dirty fix myself and this looks to be it. Although it seems the more proper and rails-y way of doing this is using the new Attributes API. [This guy did](http://til.obiefernandez.com/posts/8c31a92080-rails-5-attributes-api-jsonb-postgres-columns).

**Comment by rromanchuk** [comment]
@singjsong thanks going to give this a try, Good Enough™ version is a good start, rather get moving forward on migrating and then come back to this.

**Comment by AlxGolubev** [comment]
What's I've also noted, that there is no way to filter data with this such of jsonb strings
```ruby
[5] pry(main)> Stack::Item.last
=> #<Stack::Server:0x00007fdd5f429188
 id: 4,
 uid: "redacted,
 name: "Application Server",
 description: "",
 type: "Stack::Server",
 metadata:
  {"role"=>"Application Server" },
 render_data: {"left_position"=>286, "top_position"=>163},
 schema_id: 2
....
[6] pry(main)> Stack::Item.where('metadata @> ?', {role: 'Application Server'}.to_json)
  Stack::Item Load (0.6ms)  SELECT "st_items".* FROM "st_items" WHERE (metadata @> '{"role":"Application Server"}')
=> []
```

```pg
database=# SELECT st_items.metadata FROM st_items;
 "{\"role\":\"Application Server\"}"
doctrina_labs=# SELECT COUNT("st_items".*) FROM "st_items" WHERE (metadata @> '{"role":"Application Server"}');
 count
-------
     0
```

**Comment by archonic** [comment]
I ran into a `undefined method 'merge' for "{}":String` error on Rails 5.2 with a column with `default: {}` after migrating, but this error vanished after restarting.

**Comment by khun84** [comment]
> @Frank004 @rromanchuk changing `load` to the following seems to be Good Enough™:
> 
> ```ruby
> if hash.nil?
>   {}
> elsif hash.is_a?(Hash) && hash.empty?
>   hash
> else
>   JSON.parse(hash)
> end.with_indifferent_access
> ```
> 
> I'm just looking for a quick and dirty fix myself and this looks to be it. Although it seems the more proper and rails-y way of doing this is using the new Attributes API. [This guy did](http://til.obiefernandez.com/posts/8c31a92080-rails-5-attributes-api-jsonb-postgres-columns).

A more elaborate example of how I fix this, `Rails 5.0.6`

```ruby
class Dummy < ApplicationRecord

  store :permits, coder: JsonbCoder

end
```

```ruby
# only use this for jsonb column store coder
module JsonbCoder
  module_function

 def dump(content)
    # pg adapter will perform something like .to_json the output of this method, avoid repeating this operation as this will save invalid json object to the database
    parse_for(content)
  end

  def load(content)
    # the store accessor expects a hash
    parse_for(content)
  end

  def parse_for(content)
    return {} if content.nil?

    if content.is_a? Hash
      content
    elsif content.is_a? String
      content.empty? ? {} : JSON.parse(content) # will fail if the input is some string like 'foo', I prefer to let it fail. Otherwise just JSON.load(content) instead
    else
      content.as_json
    end
  end
end
